README.md
<p align="center">
  <a href="https://www.laplateforme.com/">
    <img src="https://www.laplateforme.com/cms/i?o=%2Fsites%2Fdefault%2Ffiles%2F2017-04%2Flogo_pdb_bsl-1.jpg" alt="Pdb logo" width="200">
  </a>
</p>

# PDB REQUESTER

[![Latest tag](https://img.shields.io/github/v/tag/Core-Techs-Git/pdb_requester?color=f87a15)](https://github.com/Core-Techs-Git/pdb_requester/tags)
![Commit since latest release](https://img.shields.io/github/commits-since/Core-Techs-Git/pdb_requester/latest?color=f87a15&sort=semver)

[![Built with TypeScript](https://img.shields.io/npm/v/typescript?color=007ACC&label=Typescript&logo=typescript)](https://github.com/microsoft/TypeScript)
[![Tested with Jest](https://img.shields.io/npm/v/jest?color=C21325&label=Jest&logo=jest)](https://github.com/facebook/jest)
[![Code Style Eslint](https://img.shields.io/npm/v/eslint?color=4B32C3&label=Eslint&logo=eslint)](https://github.com/eslint/eslint)
[![Code Style Prettier](https://img.shields.io/npm/v/prettier?color=F7B93E&label=Prettier&logo=prettier)](https://github.com/prettier/prettier)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

Part of the **[La Plateform du bâtiment](https://www.laplateforme.com/)** environment, this module is responsible for the management of the requests made to external services via a proxy (or not, according to the configuration).

## Features

Below are the features offered by **pdb_requester**:

- [x] HTTP/HTTPS requests
- [x] Requests over proxy

## How to use it ?

1. Install with `npm`

```Shell
$ npm i -S @core-techs-git/pdb_requester
```

> ⚠️
> To install this way you need to have configured `npm` to use `github package registry`. To know more you may follow [this link](https://help.github.com/en/articles/configuring-npm-for-use-with-github-package-registry#authenticating-to-github-package-registry).

<details>
  <summary>Or you can follow this quick setup</summary>
  <ul>
    <li>
      <p>Set a scoped registry access</p>
      <pre>$ npm config set @core-techs-git:registry https://npm.pkg.github.com/core-techs-git</pre>
    </li>
    <li>
      <p>Set authentication information inside your .npmrc file</p>
      <pre>//npm.pkg.github.com/:_authToken=PERSONAL-ACCESS-TOKEN</pre>
    </li>
  </ul>
  <p><b><i><a href="https://help.github.com/en/articles/creating-a-personal-access-token-for-the-command-line">PERSONAL-ACCESS-TOKEN</a></i></b> is generated in github settings.</p>
</details>

You may want to install from github url:

```Shell
$ npm i -S Core-Techs-Git/pdb_requester.git[#<commit-ish>]
```

2. Edit your configuration file _`config/dev.js`_ located at the root of your project. [config](https://github.com/lorenwest/node-config) module is used to load the right environment configurations.

```JavaScript
const config = {};
...
config.search = {
  protocol: 'https',
  host: 'search.io',
  path: '/hashcontent/orwhateveryoulike',
  defaultStoreUid: 1,
  proxy: false,    // true if requests must go through the proxy.
};
...
// An other possible config that will take advantage of baseUrl
config.bills = {
  protocol: 'https',
  host:'bills.com',
  path: 'hashcontent/', // Can be omitted
  proxy: true,
};
module.exports = config;
```

Make sure you have a _`proxy`_ key inside each top level service key in the exported object, so you may configure the proxy access.

> ⚠️
> You also need to set an environment variable to define the proxy url: `http_proxy=http(s)://host.name:port`

3. Import the module inside your code and just use it (remember that _`requester`_ is a little wrapper arround the module [Request](https://github.com/request/request))

```JavaScript
const requester = require('requester')('search');
requester.get(
  // Same parameters as node native http or https module
  // However you can pass a key "body":<string> to send in the request
  {
    url: config.search.url,
  },
  // Callback
  (err, response, body) => {
    console.log('err', err);
    console.log('response', response);
    console.log('body', body);
  });
...
const request = require('requester')('bills');
request('', // sub path like /anotherpath?with=params can be pass
(err, response, body) => {
  console.log('err', err);
  console.log('response', response);
  console.log('body', body);
});
```

## How to contribute ?

1. Import the project from [github](https://github.com/Core-Techs-Git/pdb_requester)

```Shell
$ git clone git@github.com:Core-Techs-Git/pdb_requester.git
```

2. Install dependencies with `npm`

```Shell
$ npm ci
```

That's it you're all setup and can start contributing :thumbsup:.


CHANGELOG.md
<p align="center">
<a href="https://www.laplateforme.com/">
<img src="https://www.laplateforme.com/cms/i?o=%2Fsites%2Fdefault%2Ffiles%2F2017-04%2Flogo_pdb_bsl-1.jpg"
alt="Pdb logo" width="200">
</a>
</p>

# PDB REQUESTER CHANGELOG

All notable changes to this project will be documented in this file.

##  [Release 1.2.3](https://github.com/Core-Techs-Git/pdb_requester/compare/1.2.2...1.2.3)  on 2020-08-17

![SECURITY](https://img.shields.io/static/v1?label=&style=flat&message=SECURITY&color=f39c12)

> 🔗: UPDATE DEPENDENCIES TO FIX SECURITY ISSUE [`3b67b99`](https://github.com/Core-Techs-Git/pdb_requester/commit/3b67b991a0147e129c48efc0df5881249b66c3d7)<br />

##  [Release 1.2.2](https://github.com/Core-Techs-Git/pdb_requester/compare/1.2.1...1.2.2)  on 2020-07-22

![BUGS](https://img.shields.io/static/v1?label=&style=flat&message=BUGS&color=e74c3c)

> 🐛: ALLOW HTTP TUNNELING VIA PROXY ([R-12807](https://redmine.laplateforme.com/redmine/issues/12807)) [`a814e5f`](https://github.com/Core-Techs-Git/pdb_requester/commit/a814e5f9254ee78d5a23ff1a8fd562c706be38df)<br />

##  [Release 1.2.1](https://github.com/Core-Techs-Git/pdb_requester/compare/1.2.0...1.2.1)  on 2020-03-28

![BUILD](https://img.shields.io/static/v1?label=&style=flat&message=BUILD&color=7f8c8d)

> 📦(core): UPDATE RELEASE SCRIPT [`98e2219`](https://github.com/Core-Techs-Git/pdb_requester/commit/98e221981a469b3b819a75a492b472742bc2a6be)<br />

![DOCUMENTATION](https://img.shields.io/static/v1?label=&style=flat&message=DOCUMENTATION&color=9b59b6)

> 📚: UPDATE CODE DOCUMENTATION [`077970f`](https://github.com/Core-Techs-Git/pdb_requester/commit/077970f710d99bf26eb7b580b7e92321237288d4)<br />

![BUGS](https://img.shields.io/static/v1?label=&style=flat&message=BUGS&color=e74c3c)

> 🐛(core): FIX CHANGELOG TEMPLATE [`75c7403`](https://github.com/Core-Techs-Git/pdb_requester/commit/75c740302bcc2aeec41480a65fb6695b4c8dc6c0)<br />

![SECURITY](https://img.shields.io/static/v1?label=&style=flat&message=SECURITY&color=f39c12)

> 🔗(core): FIX SECURITY VULNERABILITIE [`06d536c`](https://github.com/Core-Techs-Git/pdb_requester/commit/06d536cd0abf463e18e0428c9fb2527842f53c32)<br />

##  [Release 1.2.0](https://github.com/Core-Techs-Git/pdb_requester/compare/1.1.1...1.2.0)  on 2020-03-23

##  [Release 1.1.1](https://github.com/Core-Techs-Git/pdb_requester/compare/1.1.0...1.1.1)  on 2020-03-03

![UNKNOWN TYPE](https://img.shields.io/static/v1?label=&style=flat&message=UNKNOWN%20TYPE&color=2c3e50)

> FIX ASSIGN TO READ ONLY PROPERTY 'PATH' OF CONFIG [`e0d6621`](https://github.com/Core-Techs-Git/pdb_requester/commit/e0d66212a6a91d83c473d2ac897e70df0b043655)<br />

##  [Release 1.1.0](https://github.com/Core-Techs-Git/pdb_requester/compare/1.0.0...1.1.0)  on 2020-02-21

![UNKNOWN TYPE](https://img.shields.io/static/v1?label=&style=flat&message=UNKNOWN%20TYPE&color=2c3e50)

> WRITE CONFIGURATION SERVICE TEST [`4f6b77e`](https://github.com/Core-Techs-Git/pdb_requester/commit/4f6b77e7986b961e14b0a8b17e958a547ea90b26)<br />
> UPDATE DEPENDENCIES [`7e395dc`](https://github.com/Core-Techs-Git/pdb_requester/commit/7e395dc92b1117912b12c91455e70aa8c62b3dd4)<br />
> CHANGE TYPESCRIPT MODULE ALIAS PATH RESOLVER [`8f0995d`](https://github.com/Core-Techs-Git/pdb_requester/commit/8f0995da70bdedb1746581cc5f903b3b00076a4a)<br />
> UPDATE TESTS SET [`217daaa`](https://github.com/Core-Techs-Git/pdb_requester/commit/217daaaeda8601f9b07b90161e340eaa19ce55e9)<br />
> CHANGE CONFIGURATION SERVICE AND ERROR MANAGMENT [`c48b4e2`](https://github.com/Core-Techs-Git/pdb_requester/commit/c48b4e27dc547279abb2643cee8f34cf22dbce13)<br />
> UPDATE OLD DEPENDENCIES AND INSTALL CI ANALYSIS [`129c4ce`](https://github.com/Core-Techs-Git/pdb_requester/commit/129c4ceb2441ae02c4965ff610760a71be42ec98)<br />
> REMOVE UNNECESSARY CODE [`88e0d10`](https://github.com/Core-Techs-Git/pdb_requester/commit/88e0d105aa3b452bea27ac6025e417bada80a12d)<br />
> TEST REQUESTER SERVICE [`2135270`](https://github.com/Core-Techs-Git/pdb_requester/commit/21352702388cb72dd6038ac61e2f6215cd99ddf3)<br />
> UPDATE TESTS TO FIT NEW CONFIGURATION SETTUP [`afa0e43`](https://github.com/Core-Techs-Git/pdb_requester/commit/afa0e43acdb0b155e6c038155a4a6844773ffbf9)<br />
> TEST PDB REQUESTER MODULE ENTRY POINT [`1e4360e`](https://github.com/Core-Techs-Git/pdb_requester/commit/1e4360e91b2c4cd75545f585bf588bf5dad3c821)<br />
> SETTUP JENKINSFILE FOR CONTINUOUS INTEGRATION [`eaa4ff3`](https://github.com/Core-Techs-Git/pdb_requester/commit/eaa4ff320547af9fc192fbaa1def4d4fb45c592e)<br />
> IMPLEMENT NEW CONFIGURATION SETTUP [`cb24d8d`](https://github.com/Core-Techs-Git/pdb_requester/commit/cb24d8db8b910852e1029940a9e7b8c071d78497)<br />
> MAKE MODIFICATIONS TO FIX CI PIPELINE [`4d9b48f`](https://github.com/Core-Techs-Git/pdb_requester/commit/4d9b48f950fddd54d5856b2201580bfc9e6a80b2)<br />
> UPDATE DOCUMENTATION [`25075c7`](https://github.com/Core-Techs-Git/pdb_requester/commit/25075c7ae9a639286e30454eb31bee8236cdc68c)<br />
> UPDATE DOCUMENTATION [`4d9d1f9`](https://github.com/Core-Techs-Git/pdb_requester/commit/4d9d1f94ecdf846e255237732332c3e0b154ed9c)<br />
> ADD FIXED EVENT IN POST COMPUTATION [`0472095`](https://github.com/Core-Techs-Git/pdb_requester/commit/04720951a42ba1675e1c7d19c76d59991decc954)<br />
> IGGNORE TSCONFIG INSIDE NPM PUBLISH PACKAGE [`13e1005`](https://github.com/Core-Techs-Git/pdb_requester/commit/13e1005d66dff705fa197983ac3218d5d654e4db)<br />
> TRY AGAIN NOTIFICATION FOR REGRESSION [`cb335ea`](https://github.com/Core-Techs-Git/pdb_requester/commit/cb335ea9420000180b70b70a7d2109461a344c5b)<br />
> TRY GET BACK TO NORMAL NOTIFICATION [`ff75974`](https://github.com/Core-Techs-Git/pdb_requester/commit/ff75974dd9c9965e0c93712ae78c6d97eccaf7ce)<br />
> FIX BACK TEST ERROR TO VERIFY CI NOTIFICATIONS [`24d8ef0`](https://github.com/Core-Techs-Git/pdb_requester/commit/24d8ef01f86bcdb274613b8ee0a773f7d4630057)<br />

.versionrc.js
const build =
  '<span style="color:#ffffff;background-color:#7f8c8d;display:inline-block;padding:6px 10px;border-radius:4px;box-shadow:0 1px 3px 0 rgba(0,0,0,.1),0 1px 2px 0 rgba(0,0,0,.06);font-weight:bold;">BUILD</span>';
const documentation =
  '<span style="color:#ffffff;background-color:#9b59b6;display:inline-block;padding:6px 10px;border-radius:4px;box-shadow:0 1px 3px 0 rgba(0,0,0,.1),0 1px 2px 0 rgba(0,0,0,.06);font-weight:bold;">DOCUMENTATION</span>';
const features =
  '<span style="color:#ffffff;background-color:#2ecc71;display:inline-block;padding:6px 10px;border-radius:4px;box-shadow:0 1px 3px 0 rgba(0,0,0,.1),0 1px 2px 0 rgba(0,0,0,.06);font-weight:bold;">FEATURES</span>';
const bugs =
  '<span style="color:#ffffff;background-color:#e74c3c;display:inline-block;padding:6px 10px;border-radius:4px;box-shadow:0 1px 3px 0 rgba(0,0,0,.1),0 1px 2px 0 rgba(0,0,0,.06);font-weight:bold;">BUGS</span>';
const enhancements =
  '<span style="color:#ecf0f1;background-color:#f1c40f;display:inline-block;padding:6px 10px;border-radius:4px;box-shadow:0 1px 3px 0 rgba(0,0,0,.1),0 1px 2px 0 rgba(0,0,0,.06);font-weight:bold;">ENHANCEMENTS</span>';
const security =
  '<span style="color:#ffffff;background-color:#f39c12;display:inline-block;padding:6px 10px;border-radius:4px;box-shadow:0 1px 3px 0 rgba(0,0,0,.1),0 1px 2px 0 rgba(0,0,0,.06);font-weight:bold;">SECURITY</span>';
const unavailable =
  '<span style="color:#ffffff;background-color:#2c3e50;display:inline-block;padding:6px 10px;border-radius:4px;box-shadow:0 1px 3px 0 rgba(0,0,0,.1),0 1px 2px 0 rgba(0,0,0,.06);font-weight:bold;">UNAVAILABLE</span>';
const head = `<p align="center">
  <a href="https://www.laplateforme.com/">
    <img src="https://www.laplateforme.com/cms/i?o=%2Fsites%2Fdefault%2Ffiles%2F2017-04%2Flogo_pdb_bsl-1.jpg" alt="Pdb logo" width="200">
  </a>
</p>

# PDB REQUESTER CHANGELOG`;

module.exports = {
  header: head,
  types: [
    {
      type: '🎬',
      hidden: true,
    },
    {
      type: '🎉',
      hidden: true,
    },
    {
      type: '🔄',
      hidden: true,
    },
    {
      type: '📦',
      section: build,
    },
    {
      type: '👷',
      section: build,
    },
    {
      type: '📚',
      section: documentation,
    },
    {
      type: '🧪',
      section: security,
    },
    {
      type: '🐛',
      section: bugs,
    },
    {
      type: '🧩',
      section: features,
    },
    {
      type: '💄',
      section: features,
    },
    {
      type: '⚡',
      section: enhancements,
    },
    {
      type: '🔒',
      section: security,
    },
    {
      type: '🔊',
      section: security,
    },
    {
      type: '🔗',
      section: security,
    },
    {
      type: '♻️',
      section: enhancements,
    },
    {
      type: '🎨',
      section: enhancements,
    },
    {
      type: '🚧',
      section: unavailable,
    },
    {
      type: '⏳',
      section: unavailable,
    },
    {
      type: '📜',
      section: documentation,
    },
    {
      type: '👥',
      section: enhancements,
    },
  ],
  releaseCommitMessageFormat: '🎉: RELEASE {{currentTag}}',
  issueUrlFormat: 'https://redmine.laplateforme.com/redmine/issues/{{id}}',
  issuePrefixes: ['R-'],
  skip: {
    tag: true,
  },
};
